h2/src/main/org/h2/table/Column.java

```diff
@@ -187,6 +189,33 @@ public Value convert(CastDataProvider provider, Value v) {
         }
     }
 
+
+    /**
+     * Converts the values in a ValueRow based on the passed column info.
+     * Creates a new instance if any of the contained item must be converted. Otherwise, returns the same {@code valueRow}.
+     *
+     * @param provider the cast information provider
+     * @param columns the column info list used for the conversation
+     * @param valueRow the holder of the values
+     * @return a ValueRow which contains the converted values
+     *
+     * @see Column#convert(CastDataProvider, Value)
+     */
+    public static ValueRow convert(CastDataProvider provider, Column[] columns, ValueRow valueRow) {
+        Value[] copy = null;
+        Value[] values = valueRow.getList();
+        for (int i = values.length; --i >= 0; ) {
+            Value v = values[i];
+            Value nv = columns[i].convert(provider, v);
+            if (v != nv) {
+                if (copy == null)
+                    copy = Arrays.copyOf(values, values.length);
+                copy[i] = nv;
+            }
+        }
+        return copy == null ? valueRow : ValueRow.get(valueRow.getType(), copy);
```

Comment:
I guess `valueRow.getType()` can return wrong data type information here, so `TypeInfo.getTypeInfo(Value.ROW, 0, 0, new ExtTypeInfoRow(column))` should be used instead.

Reply:
Fixed